### PR TITLE
BGP Convergence Test for Benchmark Performance 

### PR DESCRIPTION
**Pull request by selldinesh**
### Description of PR

Summary:BGP Convergence Test for Benchmark Performance 
BGP Failover convergence for remote link failure
BGP RIB-IN Convergence
BGP RIB-IN Capacity
BGP Failover convergence for Local Link Failure
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911

### Documentation 

https://github.com/Azure/sonic-mgmt/blob/master/docs/testplan/BGP%20Convergence%20Testplan%20for%20single%20DUT.md#test-case--1--convergence-performance-when-remote-link-fails-route-withdraw

### CODE REVIEW

**Comment by wangxin** [comment]
Can you fix the testing failures related with this change? It would be great if you can fix the LGTM alerts as well.
```
==================================== ERRORS ====================================
__________ ERROR collecting ixia/bgp/test_bgp_local_link_failover.py ___________
ImportError while importing test module '/var/src/s/tests/ixia/bgp/test_bgp_local_link_failover.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
ixia/bgp/test_bgp_local_link_failover.py:2: in <module>
    from tests.common.snappi.snappi_fixtures import cvg_api
common/snappi/snappi_fixtures.py:5: in <module>
    import snappi_convergence
E   ImportError: No module named snappi_convergence
__________ ERROR collecting ixia/bgp/test_bgp_remote_link_failover.py __________
ImportError while importing test module '/var/src/s/tests/ixia/bgp/test_bgp_remote_link_failover.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
ixia/bgp/test_bgp_remote_link_failover.py:1: in <module>
    from tests.common.snappi.snappi_fixtures import cvg_api
common/snappi/snappi_fixtures.py:5: in <module>
    import snappi_convergence
E   ImportError: No module named snappi_convergence
___________ ERROR collecting ixia/bgp/test_bgp_rib_in_convergence.py ___________
ImportError while importing test module '/var/src/s/tests/ixia/bgp/test_bgp_rib_in_convergence.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
ixia/bgp/test_bgp_rib_in_convergence.py:1: in <module>
    from tests.common.snappi.snappi_fixtures import cvg_api
common/snappi/snappi_fixtures.py:5: in <module>
    import snappi_convergence
E   ImportError: No module named snappi_convergence

```

**Comment by selldinesh** [comment]
@wangxin Raised a PR https://github.com/Azure/sonic-buildimage/pull/8030 for snappi_convergence module

**Comment by selldinesh** [comment]
@wangxin can you please resolve the merge conflict, i don't have write access

**Comment by selldinesh** [comment]
Please merge this PR  ( https://github.com/Azure/sonic-buildimage/pull/8790 )for the Reboot and convergence cases to work

**Comment by yxieca** [comment]
@baiwei0427 please take a look.

File: docs/testplan/BGP-Convergence-Testplan-for-Benchmark-Performance.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+### SONiC switch as ToR
+
+![SONiC DUT as ToR ](Img/Switch_as_ToR.png)
+
+### SONiC switch as Leaf
+
+![SONiC DUT as ToR ](Img/Switch_acting_as_leaf.png)
+
+## Setup configuration
+IPv4 EBGP neighborship will be configured between SONiC DUT and directly connected test ports. Test ports inturn will simulate the ToR's and Leafs by advertising IPv4/IPv6, dual-stack routes.
```
Comment (baiwei0427) [inline docs/testplan/BGP-Convergence-Testplan-for-Benchmark-Performance.md:52]: How about BGP sessions between keysight emulated switches?

File: docs/testplan/BGP-Convergence-Testplan-for-Benchmark-Performance.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+## Test Methodology
+Following test methodologies will be used for measuring convergence. 
+* Traffic generator will be used to configure ebgp peering between chassis ports and SONiC DUT by advertising IPv4/IPv6, dual-stack routes. 
+* Receiving ports will be advertising the same VIP(virtual IP) addresses.
```
Comment (baiwei0427) [inline docs/testplan/BGP-Convergence-Testplan-for-Benchmark-Performance.md:57]: Why use the same VIP address? This is different from our data center network set up.

File: docs/testplan/BGP-Convergence-Testplan-for-Benchmark-Performance.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+## Overview
+The purpose of these tests is to test the overall convergence of a data center network by simulating multiple network devices such as ToR/Leafs and using SONiC switch DUT as one of the ToR/Leaf, closely resembling production environment.
+
+### Scope
+These tests are targeted on fully functioning SONiC system. The purpose of these tests are to measure convergence when some unexpected failures such as remote link failure, local link failure, node failure or link faults etc occur and some expected failures such as maintenance or upgrade of devices occur in the SONiC system.
+
+### Keysight Testbed
+The tests will run on following testbeds:
+* t0
```
[the remark this answers is not in the document]
Comment (selldinesh) [reply docs/testplan/BGP-Convergence-Testplan-for-Benchmark-Performance.md:38]: We are using Keysight topology for the test case, we won't be using To or T1 topology here.

File: docs/testplan/BGP-Convergence-Testplan-for-Benchmark-Performance.md (hunk 1 above)
Comment (selldinesh) [reply docs/testplan/BGP-Convergence-Testplan-for-Benchmark-Performance.md:52]: We are using eBGP peering between the DUT and the spine, in order to simulate the emulated leaf switches we are appending the AS-PATHS to the routes which are being advertised.

File: docs/testplan/BGP-Convergence-Testplan-for-Benchmark-Performance.md (hunk 2 above)
Comment (selldinesh) [reply docs/testplan/BGP-Convergence-Testplan-for-Benchmark-Performance.md:57]: In order to simulate multiple paths to the same leaf switch we have to advertise the same VIP addresses.

**Comment by wangxin** [comment]
PR test failed with below error:
```
==================================== ERRORS ====================================
_________ ERROR collecting snappi/bgp/test_bgp_local_link_failover.py __________
ImportError while importing test module '/var/src/s/tests/snappi/bgp/test_bgp_local_link_failover.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
snappi/bgp/test_bgp_local_link_failover.py:4: in <module>
    from files.bgp_convergence_helper import run_bgp_local_link_failover_test
E   ImportError: No module named bgp_convergence_helper
_________ ERROR collecting snappi/bgp/test_bgp_remote_link_failover.py _________
ImportError while importing test module '/var/src/s/tests/snappi/bgp/test_bgp_remote_link_failover.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
snappi/bgp/test_bgp_remote_link_failover.py:4: in <module>
    from files.bgp_convergence_helper import run_bgp_remote_link_failover_test
E   ImportError: No module named bgp_convergence_helper
___________ ERROR collecting snappi/bgp/test_bgp_rib_in_capacity.py ____________
ImportError while importing test module '/var/src/s/tests/snappi/bgp/test_bgp_rib_in_capacity.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
snappi/bgp/test_bgp_rib_in_capacity.py:4: in <module>
    from files.bgp_convergence_helper import run_RIB_IN_capacity_test
E   ImportError: No module named bgp_convergence_helper
__________ ERROR collecting snappi/bgp/test_bgp_rib_in_convergence.py __________
ImportError while importing test module '/var/src/s/tests/snappi/bgp/test_bgp_rib_in_convergence.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
snappi/bgp/test_bgp_rib_in_convergence.py:4: in <module>
    from files.bgp_convergence_helper import run_rib_in_convergence_test
E   ImportError: No module named bgp_convergence_helper

```

@selldinesh  Can you fix the issue?

**Comment by selldinesh** [comment]
@wangxin 

bgp_convergence_helper module is part of the PR 
bgp_convergence_helper.py  file

**Comment by wangxin** [comment]
PR test still failed with same issue:
```
==================================== ERRORS ====================================
_________ ERROR collecting snappi/bgp/test_bgp_local_link_failover.py __________
ImportError while importing test module '/var/src/s/tests/snappi/bgp/test_bgp_local_link_failover.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
snappi/bgp/test_bgp_local_link_failover.py:4: in <module>
    from files.bgp_convergence_helper import run_bgp_local_link_failover_test
E   ImportError: No module named bgp_convergence_helper
_________ ERROR collecting snappi/bgp/test_bgp_remote_link_failover.py _________
ImportError while importing test module '/var/src/s/tests/snappi/bgp/test_bgp_remote_link_failover.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
snappi/bgp/test_bgp_remote_link_failover.py:4: in <module>
    from files.bgp_convergence_helper import run_bgp_remote_link_failover_test
E   ImportError: No module named bgp_convergence_helper
___________ ERROR collecting snappi/bgp/test_bgp_rib_in_capacity.py ____________
ImportError while importing test module '/var/src/s/tests/snappi/bgp/test_bgp_rib_in_capacity.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
snappi/bgp/test_bgp_rib_in_capacity.py:4: in <module>
    from files.bgp_convergence_helper import run_RIB_IN_capacity_test
E   ImportError: No module named bgp_convergence_helper
__________ ERROR collecting snappi/bgp/test_bgp_rib_in_convergence.py __________
ImportError while importing test module '/var/src/s/tests/snappi/bgp/test_bgp_rib_in_convergence.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
snappi/bgp/test_bgp_rib_in_convergence.py:4: in <module>
    from files.bgp_convergence_helper import run_rib_in_convergence_test
E   ImportError: No module named bgp_convergence_helper
```
Probably something wrong with the way importing the dependency?

Can you run test run with your change on a t0 VS setup using command like below:
```
./run_tests.sh -i veos_vtb -d vlab-01 -n vms-kvm-t0 -f vtestbed.csv -k debug -l warning -m individual -q 1 -a False -O -r -E -c '    monit/test_monit_status.py     platform_tests/test_advanced_reboot.py     test_interfaces.py     arp/test_arp_dualtor.py     bgp/test_bgp_fact.py     bgp/test_bgp_gr_helper.py::test_bgp_gr_helper_routes_perserved     bgp/test_bgp_speaker.py     bgp/test_bgp_update_timer.py     cacl/test_ebtables_application.py     cacl/test_cacl_application.py     cacl/test_cacl_function.py     dhcp_relay/test_dhcp_relay.py     dhcp_relay/test_dhcpv6_relay.py     lldp/test_lldp.py     ntp/test_ntp.py     pc/test_po_cleanup.py     pc/test_po_update.py     route/test_default_route.py     route/test_static_route.py     arp/test_neighbor_mac.py     arp/test_neighbor_mac_noptf.py     snmp/test_snmp_cpu.py     snmp/test_snmp_interfaces.py     snmp/test_snmp_lldp.py     snmp/test_snmp_pfc_counters.py     snmp/test_snmp_queue.py     snmp/test_snmp_loopback.py     snmp/test_snmp_default_route.py     ssh/test_ssh_stress.py     ssh/test_ssh_ciphers.py     syslog/test_syslog.py     tacacs/test_rw_user.py     tacacs/test_ro_user.py     tacacs/test_ro_disk.py     tacacs/test_jit_user.py     telemetry/test_telemetry.py     test_features.py     test_procdockerstatsd.py     iface_namingmode/test_iface_namingmode.py     platform_tests/test_cpu_memory_usage.py     bgp/test_bgpmon.py     container_checker/test_container_checker.py     process_monitoring/test_critical_process_monitoring.py     system_health/test_system_status.py' -p logs/1vlan
```

**Comment by selldinesh** [comment]
added __init__.py file under snappi/bgp path.
Can you try running again. for me its working

**Comment by wangxin** [comment]
@selldinesh Now we have a different error.

It looks like this test depends on python package `statistics` in sonic-mgmt docker. Can you add this package to the sonic-mgmt docker image by submitting a PR to the sonic-buildimage repo?

```
==================================== ERRORS ====================================
_________ ERROR collecting snappi/bgp/test_bgp_local_link_failover.py __________
ImportError while importing test module '/var/src/s/tests/snappi/bgp/test_bgp_local_link_failover.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
snappi/bgp/test_bgp_local_link_failover.py:4: in <module>
    from files.bgp_convergence_helper import run_bgp_local_link_failover_test
snappi/bgp/files/bgp_convergence_helper.py:2: in <module>
    from statistics import mean
E   ImportError: No module named statistics
_________ ERROR collecting snappi/bgp/test_bgp_remote_link_failover.py _________
ImportError while importing test module '/var/src/s/tests/snappi/bgp/test_bgp_remote_link_failover.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
snappi/bgp/test_bgp_remote_link_failover.py:4: in <module>
    from files.bgp_convergence_helper import run_bgp_remote_link_failover_test
snappi/bgp/files/bgp_convergence_helper.py:2: in <module>
    from statistics import mean
E   ImportError: No module named statistics
___________ ERROR collecting snappi/bgp/test_bgp_rib_in_capacity.py ____________
ImportError while importing test module '/var/src/s/tests/snappi/bgp/test_bgp_rib_in_capacity.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
snappi/bgp/test_bgp_rib_in_capacity.py:4: in <module>
    from files.bgp_convergence_helper import run_RIB_IN_capacity_test
snappi/bgp/files/bgp_convergence_helper.py:2: in <module>
    from statistics import mean
E   ImportError: No module named statistics
__________ ERROR collecting snappi/bgp/test_bgp_rib_in_convergence.py __________
ImportError while importing test module '/var/src/s/tests/snappi/bgp/test_bgp_rib_in_convergence.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
snappi/bgp/test_bgp_rib_in_convergence.py:4: in <module>
    from files.bgp_convergence_helper import run_rib_in_convergence_test
snappi/bgp/files/bgp_convergence_helper.py:2: in <module>
    from statistics import mean
E   ImportError: No module named statistics

```

**Comment by selldinesh** [comment]
@wangxin Created A PR https://github.com/Azure/sonic-buildimage/pull/9682/ for the ImportError: No module named statistics issue

**Comment by selldinesh** [comment]
@wangxin  Azure/sonic-buildimage#9682 got merged, can you please update the status of the bgp case run

**Comment by selldinesh** [comment]
@wangxin Any update on the run?

**Comment by selldinesh** [comment]
@wangxin can you please merge https://github.com/Azure/sonic-buildimage/pull/9682.
This needs atleast one reviewer to approve, when this is fixed the statisticss import issue will be fixed

**Comment by wangxin** [comment]
> 

PR https://github.com/Azure/sonic-buildimage/pull/9682 has been merged, lets wait for a new sonic-mgmt docker image.

**Comment by selldinesh** [comment]
@wangxin I have updated the BGP scripts for the latest snappi version and raised a PR https://github.com/Azure/sonic-buildimage/pull/9932 for snappi version 0.7.12 , can you please merge the PR 9932 and create a new sonic-mgmt docker image, after which the pipeline may pass

**Comment by selldinesh** [comment]
@wangxin The checks were successful before, now i see some checks are not successful, the results shows the test setup failure. Can  this be resolved and merged

**Comment by wangxin** [comment]
@selldinesh I am not able to merge this PR until all tests passed. Lets give it another try.

**Review by baiwei0427 — DISMISSED** [review]

File: tests/snappi/bgp/files/bgp_convergence_helper.py [hunk 4]
```diff
@@ -941,5 +941,5 @@ def cleanup_config(duthost):
     duthost.command("sudo cp {} {}".format("/etc/sonic/config_db_backup.json","/etc/sonic/config_db.json"))
     duthost.shell("sudo config reload -y \n")
     logger.info("Wait until all critical services are fully started")
-    pytest_assert(wait_until(360, 10, 1, duthost.critical_services_fully_started), "Not all critical services are fully started")
+    pytest_assert(wait_until(360, 10, duthost.critical_services_fully_started), "Not all critical services are fully started")
```
Comment (wangxin) [inline tests/snappi/bgp/files/bgp_convergence_helper.py:944]: I am afraid this change won't work. 

The `wait_until` function has been improved to support initial delay. Unfortunately the change is not backward compatible. The author has made a swept change. But all the open PRs still suffer from this issue.
This is the new signature of the `wait_until` function: https://github.com/Azure/sonic-mgmt/blob/master/tests/common/utilities.py#L82

For the `wait_until` to work in your local setup, can you do a rebase to the latest master branch and make a force push

**Review by wangxin — CHANGES_REQUESTED** [review]

File: tests/snappi/bgp/files/bgp_convergence_helper.py (hunk 4 above)
Comment (selldinesh) [reply tests/snappi/bgp/files/bgp_convergence_helper.py]: @wangxin i have reverted the changes back for the wait_until function

**Review by wangxin — APPROVED** [review]

**Pull request merged**